codice/ddf
# DDF-2031 Adds a two-second delay on failed login to prevent automated password cracking

#### What does this PR do?
Adds a delay mechanism on failed login attempts.

Also removes synchronization from `LoginFilter` and fixes `AbstractStsRealm` so it no longer shares an `STSClient` between threads.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@tbatie 
@roelens8 

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@kcwire
@stustison

#### How should this be tested?
A full build will exercise the unit tests. Further manual testing is probably best managed by attaching a remote debugger to a running DDF instance and confirming that the delay code is invoked. Additionally, any attempt to login through the STS with incorrect credentials will be timed out so scripting an attack-like scenario should demonstrate that the system slows down the operation of the script while still allowing correct logins to pass.

#### Any background context you want to provide?
N/A

#### What are the relevant tickets?
DDF-2031

#### Screenshots (if appropriate)
N/A

#### Checklist:
